dags/mlflow-multimodel-dag.py

```diff
@@ -0,0 +1,274 @@
+            ['log_reg', LogisticRegression(max_iter=500)]
+        ]
+
+        tasks = []
+
+        for model in models:
+            @task(task_id=model[0])
+            def train(df: pd.DataFrame, model_type=model[0],model=model[1], grid_params=model_grid_params[model[0]], **kwargs):
+
+                y = df['never_married']
+                X = df.drop(columns=['never_married'])
+
+
+                X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=55, stratify=y)
+                
+                mlflow.set_tracking_uri('http://host.docker.internal:5000')
+                try:
+                    # Creating an experiment 
+                    mlflow.create_experiment('census_prediction')
+                except:
+                    pass
+                # Setting the environment with the created experiment
+                mlflow.set_experiment('census_prediction')
+
+                grid_search = GridSearchCV(model, param_grid=grid_params, verbose=1, cv=5, n_jobs=-1)
+
+
+                mlflow.sklearn.autolog()
+
+                with mlflow.start_run(run_name=f'{model_type}_{kwargs["run_id"]}'):
+
+                    logging.info('Performing Gridsearch')
+                    grid_search.fit(X_train, y_train)
+
+                    logging.info(f'Best Parameters\n{grid_search.best_params_}')
+                    best_params = grid_search.best_params_
+
+
+                    if model_type == 'lgbm':
+
+                        mlflow.sklearn.autolog()
+
+                        train_set = lgb.Dataset(X_train, label=y_train)
+                        test_set = lgb.Dataset(X_test, label=y_test)
+
+                        best_params['metric'] = ['auc', 'binary_logloss']
+
+                        logging.info('Training model with best parameters')
+                        clf = lgb.train(
+                            train_set=train_set,
+                            valid_sets=[train_set, test_set],
+                            valid_names=['train', 'validation'],
+                            params=best_params,
+                            early_stopping_rounds=5
+                        )
+                        
+                    else:
+                        logging.info('Training model with best parameters')
+                        clf = LogisticRegression(penalty=best_params['penalty'], C=best_params['C'], solver=best_params['solver']).fit(X_train, y_train)
+
+                    y_pred_class = test(clf, X_test, model_type)
+
+                    # Log Classfication Report, Confustion Matrix, and ROC Curve
+                    log_all_eval_metrics(y_test, y_pred_class)
+
+            tasks.append(train(features))
+        return tasks
+
+    df = load_data()
+    clean_data = preprocessing(df)
+    features = feature_engineering(clean_data)
+    grid_search_cv(features)
+
+    
+dag = mlflow_multimodel_example()
```

Comment:
So there's so much this dag has in common with the previous one (as it should), I am wondering if it won't be cleaner to factorize them. That would be my preference, but it's your call ofc. 

In the Registry, will these show up as separate entries?

Reply:
If we keep as is, my comments from the previous dag apply here too.